### PR TITLE
fix: stop jobs getting stranded at 'queued' after Kubernetes reports them complete

### PR DESCRIPTION
**Pull request by matthewberry**
## The bug

Jobs that Kubernetes reports as complete keep coming back as `queued` from the API, and never catch up.

`GET /{job_type}/jobs/{job_id}` returns the DB `phase` column and nothing else, and the only writer of that column is `KubeEventWatcher`. A k8s watch replays nothing older than the `resourceVersion` it reconnects at, and nothing reconciled after the fact — so any event the watcher missed was lost **permanently**, leaving the row at whatever phase it last had (`queued` if no event for it was ever processed).

Three things conspired to lose those events:

**1. The DB row was written after the Kubernetes Job.** `app/routers/job.py:307` created the k8s Job; the row wasn't committed until `:342`. The watcher receives the `ADDED` event in between and its lookup by `jobId` returns `None`.

**2. That `None` crashed the watcher and tore down the watch stream.** `kubejob_service.py:319-322` logged `"None" was encountered`, skipped the DB write, then still called `send_notification_email(job_id, job_type, None, new_phase)` — which dereferences `updated_job.job_id` while building the results URL, raising `AttributeError`. The outer handler caught it as if it were a connection fault: stream closed, `sleep(2)`, reconnect. On reconnect the code re-lists **only to read `resource_version`** and throws the returned Jobs away, so every Job transition in that window is gone for good.

So each submission opened a blind window at the busiest moment. A job whose completion landed in one was `Complete` in k8s and `queued` in the DB forever — and short jobs (OED/CLEANDB are explicitly "very fast") can fit their whole lifecycle inside it. Jobs created in the same window are the worst case: their `ADDED` event *is* the one that crashed the watcher, so nothing for them was ever processed.

**3. Completion detection also missed real completions.** `kubejob_service.py:297` accepted success only as `conditions[0].type == 'SuccessCriteriaMet'`. `Complete` — the condition every successful Job gets, and the only one on clusters older than 1.31 — was never checked, and the fixed index made the test order-dependent. A non-matching terminal event fell through to an `else` that just logged `>> Skipped job update`.

## The fix

- **`routers/job.py`** — create and commit the Job row **before** `create_job()`. If creation then fails, mark that row `error` instead of leaving a job that will never run at `queued`; guarded on `created_new_job` so a duplicate POST cannot knock an existing job out of its legitimate phase.
- **`kubejob_service.py`** — new `_derive_phase()` reads terminal state from the whole condition list (`Complete`/`SuccessCriteriaMet` to completed, `Failed`/`FailureTarget` to error) with `succeeded`/`failed` vs `spec.completions` as a fallback.
- **`kubejob_service.py`** — extracted `_reconcile_job_phase()`, now also run over the listed Jobs on every (re)connect, so anything missed is caught up rather than stranded. The DB write stays guarded on an actual phase change, and emails stay idempotent via the existing MinIO marker. The missing-row case returns before the email call.
- **`kubejob_service.py`** — per-event and per-job failures are isolated, so one bad Job can no longer cost the stream; the watch is bounded at 600s so the reconcile pass runs on a known cadence (completed Jobs are still listed within `ttlSecondsAfterFinished`, 12h).
- **`kubejob_service.py`** — `getattr(e, 'status', None)`: `requests.HTTPError` has no `.status`, so the unguarded read raised *inside* the exception handler, escaped `run()` and killed the watcher thread outright — after which every job stays `queued`, and nothing checks `is_alive()`.
- **`main.py`** — dropped `watcher.run()` from the lifespan. The thread is already started in `__init__`; calling the thread body inline would block the event loop at startup. It is inert today only because FastAPI 0.89.0 predates `lifespan=` support and swallows it into `**extra`.

## Testing

`_derive_phase` was exercised against 15 real `V1JobStatus` shapes, including the two the old logic got wrong (`[Complete]` alone; `Complete` not at index 0) and the case that previously fell through silently (DeadlineExceeded with `failed=0`). Not yet run against a live cluster — I have no working `mmli1` credentials.

Worth checking in the deployed logs before/after: `grep -E 'None" was encountered|Skipped job update|reconnecting'`. A total absence of `Updating job phase` after some timestamp would mean the thread had already died via the `e.status` path.

## Notes for review

- Existing stranded rows will not fix themselves on deploy if their Jobs have aged out of the 12h TTL — those need a manual phase update.
- The reconcile pass now calls `send_notification_email` for terminal jobs every ~10 min; it does one MinIO stat per completed job that has an email set, and the marker keeps it from re-sending.
- PR #105 carries earlier versions of the `_reconcile_job_phase` extraction and the `None` guard on its own branch; expect an overlapping patch when that merges.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by bodom0015** [comment]
Thank you, Matt! This is definitely in the right neighborhood, and could be an odd edge case that we've missed for a looong time 👏 

I'm not sure why this PR didn't trigger an autobuild 🤔🙏 I can try and trigger one manually tho so we can test it out!

EDIT: I was able to trigger a Docker autobuild by fixing the merge conflicts 🙏 still unclear why this did not trigger when the PR was created. It is also unclear why there were so many changes pulled in here? Some of these changes are from the mmli2 migration back in January. Just asking for clarification: @matthewberry is your main branch up-to-date locally? If not, you may need to `git checkout main` and `git pull origin main`? 🤔 verbose log output below from fixing the merge conflicts:
```
lambert8@NCSA-P10H04738 mmli-backend % git status
On branch mjberry/fix-stuck-queued-jobs
Your branch is up to date with 'origin/mjberry/fix-stuck-queued-jobs'.

You have unmerged paths.
  (fix conflicts and run "git commit")
  (use "git merge --abort" to abort the merge)

Changes to be committed:
	modified:   .github/workflows/docker.yml
	modified:   .gitignore
	modified:   Dockerfile
	modified:   app/cfg/config.yaml
	modified:   app/config.py
	new file:   app/coordinator.py
	modified:   app/main.py
	modified:   app/models/enums.py
	modified:   app/postjob.py
	modified:   app/routers/chemscraper.py
	modified:   app/routers/files.py
	new file:   app/routers/reactionminer.py
	modified:   app/routers/shared.py
	modified:   app/services/clean_service.py
	new file:   app/services/ezspecificity_service.py
	modified:   app/services/kubejob_service.py
	modified:   app/services/minio_service.py
	modified:   app/services/oed_service.py
	modified:   app/services/rdkit_service.py
	new file:   app/services/reactionminer_search/NOTES.txt
	new file:   app/services/reactionminer_search/__init__.py
	new file:   app/services/reactionminer_search/core.py
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_full/Par_Imgs.tsv
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_full/Par_SMI.tsv
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_full/Par_SMI_Names.tsv
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_full/Pars.tsv
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_full/Pars2SMI.tsv
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_reactions_only/Par_Imgs.tsv
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_reactions_only/Par_SMI.tsv
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_reactions_only/Par_SMI_Names.tsv
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_reactions_only/Pars.tsv
	new file:   app/services/reactionminer_search/indexing_data_suzuki/tsv_files_reactions_only/Pars2SMI.tsv
	modified:   app/services/reactionminer_service.py
	modified:   app/services/shared.py
	new file:   app/services/simplefold_service.py
	modified:   app/services/templates/job.tpl.yaml
	modified:   chart/Chart.lock
	modified:   chart/Chart.yaml
	renamed:    chart/weights.pvc.yaml -> chart/mmli-shared-job-data.pvc.yaml
	modified:   chart/pvc-explorer.pod.yaml
	modified:   chart/templates/deployment.yaml
	modified:   chart/templates/service.yaml
	modified:   chart/values.local.yaml
	new file:   chart/values.mmli2.prod.yaml
	new file:   chart/values.mmli2.staging.yaml
	modified:   chart/values.prod.yaml
	modified:   chart/values.staging.yaml
	modified:   chart/values.yaml

Unmerged paths:
  (use "git add <file>..." to mark resolution)
	both modified:   app/routers/job.py

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	... a bunch of my local junk that can be ignored ... 
	
	<---   At this point, I went ahead and manually fixed the app/routers/job.py file    --->
	<---   The conflict was trivial (one was empty, other had added code), so I committed this and pushed it    --->
	
lambert8@NCSA-P10H04738 mmli-backend % git add app/routers/job.py
lambert8@NCSA-P10H04738 mmli-backend % git commit -m 'fix: merge conflicts?'
[mjberry/fix-stuck-queued-jobs 5070466] fix: merge conflicts?
lambert8@NCSA-P10H04738 mmli-backend % git push origin mjberry/fix-stuck-queued-jobs
Enumerating objects: 22, done.
Counting objects: 100% (22/22), done.
Delta compression using up to 12 threads
Compressing objects: 100% (8/8), done.
Writing objects: 100% (8/8), 2.32 KiB | 2.32 MiB/s, done.
Total 8 (delta 7), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (7/7), completed with 7 local objects.
remote: 
remote: GitHub found 5 vulnerabilities on moleculemaker/mmli-backend's default branch (5 moderate). To find out more, visit:
remote:      https://github.com/moleculemaker/mmli-backend/security/dependabot
remote: 
To https://github.com/moleculemaker/mmli-backend
   cb42758..5070466  mjberry/fix-stuck-queued-jobs -> mjberry/fix-stuck-queued-jobs
```

**Comment by matthewberry** [comment]
Found while stacking #111 (characterization tests) on this branch — not a functional bug, but worth cleaning up before merge.

## `create_job` has an unreachable duplicate of its own DB-write block

The merge commit `5070466 "fix: merge conflicts?"` kept **both** copies of the row-creation logic: the relocated one this PR adds, and the original tail it was moved from.

`app/routers/job.py`:

- **L440–478** — the new block. `created_new_job = not db_job`, then creates the row *before* `create_job()`. This is the fix.
- **L490–497** — `if created_new_job: return 201`
- **L508–538** — the original tail: `if not db_job:` → build `Job(...)`, commit, `return 201`

**L508–538 can never execute.** Reaching L503 requires falling out of the `if job_type in JobTypes:` branch without returning, which only happens when `created_new_job` is `False` — L490 returns in every other case. And `created_new_job is False` implies `db_job` is truthy, so `if not db_job:` at L508 is always `False`.

Only the `else:` at L539 is live. `user_agent` at L503 is likewise dead — its sole consumer is the `Job(...)` constructor inside the unreachable block.

## Suggested fix

Delete L503–538, leaving the live path as:

```python
    # db_job already existed: this POST is an update, not a creation
    db_job.job_info = job_info
    await db.merge(db_job)
    await db.commit()
    await db.refresh(db_job)

    return JSONResponse(status_code=status.HTTP_200_OK, content={...})
```

The `if not db_job:` / `else:` pair collapses to just the former `else` body, since the branch condition is now a constant.

## Why it's worth doing here

The two copies differ in a way that invites a wrong future edit: the surviving one has no `created_new_job` guard and no error handling around `create_job()`. Anyone reading `create_job` top-to-bottom hits the second block and reasonably concludes row creation still happens after the Kubernetes call — which is exactly the ordering this PR exists to fix.

## Verification

40 characterization tests in #111 pass identically against this branch and against `main`, so this PR does not change any observable response — consistent with the duplicate being dead rather than doing anything.

## Not asking you to fix this here

The surviving `else` branch at L539–545 overwrites the stored `job_info` on a duplicate POST, and the 200 response at L546 echoes back the existing job's `email`. That predates this PR and is being handled in a follow-up in the same stack — flagging it only so it doesn't look like an oversight.

**Review by bodom0015 — APPROVED** [review]
LGTM :+1: any extra error-handling seems like a good thing here, given the number of opportunities for surprise edge cases 😅 and sorry if I created additional problems from a botched merge

Do we need to resolve the conflicts here? I apparently am not reliable for that when too much is changing too fast 😅

File: app/routers/job.py [hunk 1]
```diff
@@ -440,8 +477,24 @@ async def create_job(
         except Exception as ex:
```
Comment (matthewberry) [inline app/routers/job.py]: **1. HIGH — the new "mark the row `error` if creation fails" handler is dead for the dominant failure mode.**

`kubejob_service.create_job()` catches `ApiException` itself (`app/services/kubejob_service.py:719-724`) and *returns* `{'status': STATUS_ERROR, 'message': msg}` instead of raising; this router ignores the return value.

So a quota/LimitRange rejection, an RBAC `Forbidden`, a 409 `AlreadyExists` on a duplicate submit, or an invalid rendered spec all leave the row committed at `app/routers/job.py:462-464` sitting at `queued` forever while the client gets `201 CREATED` — precisely the outcome this PR sets out to eliminate. Only non-`ApiException` errors (jinja/yaml/template) reach this handler.

Fix: inspect the returned dict's `status`/`message` rather than relying on an exception.

**Review by matthewberry — COMMENTED** [review]
Review of the diff (`origin/main...origin/mjberry/fix-stuck-queued-jobs`, 4 commits). Seven findings; six are posted inline below, one could not be anchored to the diff and is reproduced here, plus scope notes.

---

**6. `app/routers/job.py:508` — LOW.** (Not part of the diff, so it cannot take an inline comment.) Dead duplicated row constructor. Reading every line from 440 to 508: every `created_new_job` path returns at line 490 and the `else` at 498 raises, so reaching 503 requires `db_job` truthy and `if not db_job:` at 508 can never be true. Lines 503-538 (including the second `user_agent = ''` at 504) are unreachable, leaving two identical 25-line `Job(...)` constructors that will silently drift apart.

---

### Scope of this review

**Did not investigate:**
- The EZSpec subjob `job_id` reassignment (`EZSPEC_UNIDOCK` / `EZSPEC_INFERENCE` rebind `job_id` from `job_info` after the duplicate-job lookup at the top of `create_job`) and how it interacts with the moved insert — beyond confirming the new block is byte-identical to the one it replaces.
- Connection-pool behavior of opening one `Session(self.engine)` per job across a full namespace listing in the new reconcile sweep.

**Looked at and did not find a problem with** — this is what I read, not a guarantee the state space was explored:
- `_derive_phase` (`kubejob_service.py:238-267`): `backoffLimit=0` is hard-coded at line 666, so `failed > 0 -> ERROR` is sound; `status is None`, `conditions is None`, and the `succeeded`/`completions` fallback each reduce to the old semantics or strictly improve on them.
- The new per-event `try/except` and watch `ERROR`/`BOOKMARK` frames: those events keep `event['object']` as a plain dict, so `event['object'].metadata` at line 383 raises *outside* the inner try and still reaches the outer handler — unchanged from `main`. `resource_version` is re-read from a fresh list each iteration, so a 410 loop self-heals.
- `main.py`: `is_alive()` exists (`kubejob_service.py:406`); `requirements.txt:1` pins `fastapi[all]==0.89.0`, which predates `lifespan=`, so the claim that the removed `watcher.run()` was inert checks out.
- `JobStatus.ERROR` import, `traceback` import, and `Session` / `self.engine` availability at each new call site.

File: app/routers/job.py [hunk 2]
```diff
@@ -440,8 +477,24 @@ async def create_job(
...
             log.error("Failed to create Job: " + str(ex))
             log.error(traceback.format_exc())
+            # The row we just wrote is already visible to clients, and nothing will ever run
+            # to move it along - fail it here rather than leaving it polling 'queued' forever.
+            # Only ever touch a row this request created: a duplicate POST for an existing
+            # job_id must not knock that job out of whatever phase it legitimately reached.
+            if created_new_job:
+                db_job.phase = JobStatus.ERROR
```
Comment (matthewberry) [inline app/routers/job.py]: **7. LOW — the error path sets `phase` but not `time_end`.**

Every other phase writer in the codebase sets the timestamp alongside the phase: `app/models/sqlmodel/db.py:48-56`, `app/services/chemscraper_service.py:42-48`, `app/services/aceretro_service.py:17`, `app/services/reactionminer_service.py:19`. A job that fails at submission will report `time_end = 0`.

File: app/services/kubejob_service.py [hunk 3]
```diff
@@ -235,6 +235,89 @@ def mark_email_as_sent(self, job_type, job_id, success):
...
+                return
+
+            if updated_job.phase != new_phase:
+                self.logger.debug('Updating job phase: %s -> %s' % (job_id, new_phase))
+                updated_job.phase = new_phase
+                session.add(updated_job)
+                session.commit()
+                session.flush()
+
+            self.send_notification_email(job_id, job_type, updated_job, new_phase)
```
Comment (matthewberry) [inline app/services/kubejob_service.py]: **2. HIGH — duplicate notification-email amplification.**

`send_notification_email` sits *outside* the `phase != new_phase` guard, and the new reconcile sweep (lines 337, 355) runs it over every Job in the namespace at least every 600 s.

The sole idempotency gate is `should_send_email` -> `MinIOService.check_file_exists`, which returns **`False` for every `S3Error` except `NoSuchKey`** (`app/services/minio_service.py:101-106`) — `NoSuchBucket`, connection refused, and auth failures all read as "not sent yet".

During a MinIO outage, every terminal job with an email address inside the 12 h `ttlSecondsAfterFinished` window (`app/cfg/config.yaml:408`, `43200`) gets a fresh "your result is ready" mail every ~10 minutes, and `mark_email_as_sent`'s upload fails too, so no marker is ever laid down. Before this PR the email path only ran on an actual watch event for one job, so a MinIO blip could duplicate at most one message.

The docstring at lines 276-278 asserts this is idempotent; that assertion is the thing under review, and it does not hold.

Fix: move the call inside the change guard and/or treat a non-`NoSuchKey` `S3Error` as "already sent".

File: app/services/kubejob_service.py [hunk 4]
```diff
@@ -235,6 +235,89 @@ def mark_email_as_sent(self, job_type, job_id, success):
...
+            if updated_job is None:
+                # The API creates the DB row before the Kubernetes Job, so this should only
+                # happen for a Job whose row was deleted. Bail out before the email: passing
+                # None into send_notification_email raises, and an exception here used to
+                # tear down the watch stream (losing every event in the reconnect gap).
+                self.logger.warning(f'"None" was encountered when fetching Job: {job_id}')
+                self.logger.warning('Skipping database update...')
+                return
+
+            if updated_job.phase != new_phase:
```
Comment (matthewberry) [inline app/services/kubejob_service.py]: **3. MEDIUM — the sweep makes the watcher deterministically clobber phases written by application code.**

`ChemScraperService.update_job_phase` sets `completed` from the `/chemscraper/run` background task (`app/services/chemscraper_service.py:287`), and the same `job_id` also has a Kubernetes Job (`app/cfg/config.yaml:72-73`, `moleculemaker/chemscraper-job`).

If that k8s Job is still Pending/Running (or failed) when the app writes `completed`, the next sweep derives `processing`/`error` and writes it back — and because the sweep only writes on a change, the row then stays wrong permanently and the frontend regresses to a spinner. Previously this required a k8s event to arrive after the app write; now it is guaranteed within 10 minutes.

Suggest refusing to downgrade a terminal DB phase to `processing`.

**Not reproduced against a live cluster.** The two writers and the k8s Job config are confirmed by reading; the timing overlap is not.

File: app/services/kubejob_service.py [hunk 5]
```diff
@@ -115,7 +115,7 @@ class KubeEventWatcher:
 
     def __init__(self):
         self.logger = log
-        #self.logger.setLevel('DEBUG')
+        self.logger.setLevel('DEBUG')
```
Comment (matthewberry) [inline app/services/kubejob_service.py]: **5. LOW — debug leftover.**

`self.logger.setLevel('DEBUG')` was deliberately commented out on `main` and is now live, pinning the `services.kubejob_service` logger to DEBUG regardless of `LOGLEVEL`. Combined with the new per-job reconcile sweep this *raises* log volume, working against the intent of the `db.py` `echo` change in the same PR. Reads as a leftover from commit `088648e`.

File: app/models/sqlmodel/db.py [hunk 6]
```diff
@@ -17,7 +17,12 @@
 
 # app_secrets['db']['url'] = "postgresql://user:password@postgresserver/db"
 def create_db_engine():
-    return AsyncEngine(create_engine(SQLALCHEMY_DATABASE_URL, echo=True, future=True))
+    return AsyncEngine(
+        create_engine(
+            SQLALCHEMY_DATABASE_URL,
+            echo=(True if SERVER_LOGLEVEL == 'DEBUG' else False),
```
Comment (matthewberry) [inline app/models/sqlmodel/db.py]: **4. LOW — this is a no-op as shipped.**

`app/cfg/config.yaml:3` sets `loglevel: "DEBUG"`, and `git grep -in 'loglevel' -- chart` returns only `chart/values.yaml:84:    loglevel: "DEBUG"` — no override in `values.prod.yaml`, `values.staging.yaml`, `values.mmli2.prod.yaml`, `values.mmli2.staging.yaml`, or the templates. So `echo` stays `True` in every deployed environment and the "turn down sqlalchemy transactions" commit changes nothing.

It also re-derives, case-sensitively, the `DEBUG` flag that `config.py:58` already exports as `SERVER_LOGLEVEL.lower() == 'debug'` — so `LOGLEVEL=debug` now behaves differently here than everywhere else in the app. Suggest `from config import DEBUG`.

File: app/services/kubejob_service.py (hunk 4 above)
Comment (matthewberry) [reply app/services/kubejob_service.py]: Not reproducible — the premise does not hold.

`ChemScraperService.update_job_phase` has exactly one caller chain: the background task started by `POST /chemscraper/analyze` (`app/routers/chemscraper.py:66` → `runChemscraperOnDocument`). That endpoint is deprecated, it builds the `Job` row itself, and it calls the external ChemScraper HTTP API — it never creates a Kubernetes Job. The sweep only ever iterates `list_namespaced_job`, so it cannot see that `job_id` at all.

```
$ git grep -n 'runChemscraperOnDocument\|update_job_phase' origin/main -- app | grep -v 'def update_job_phase'
origin/main:app/routers/chemscraper.py:66:    background_tasks.add_task(chemscraperService.runChemscraperOnDocument, ...)
origin/main:app/services/chemscraper_service.py:256:        await self.update_job_phase(db_job, JobStatus.PROCESSING)
origin/main:app/services/chemscraper_service.py:262:        await self.update_job_phase(db_job, JobStatus.ERROR)
origin/main:app/services/chemscraper_service.py:288:        await self.update_job_phase(db_job, JobStatus.COMPLETED)
origin/main:app/services/chemscraper_service.py:299:        await self.update_job_phase(db_job, JobStatus.ERROR)
```

The `kubernetes_jobs.chemscraper` entry the finding cites is used by `POST /chemscraper/jobs`, which goes through the generic router and never touches `ChemScraperService`. The two writers are on disjoint paths.

The general shape — the watcher overwriting a terminal phase the application set — is real, and `main` now guards the one case that actually occurs: `_reconcile_job_phase` returns early for `JobStatus.CANCELED`, because deleting a Job produces events that derive to `processing`. The remaining way in is a client calling `PUT`/`PATCH /{job_type}/jobs/{job_id}/{run_id}` directly; nothing in the repo does, and `jobCompleteApiUrl` is rendered into the template but never referenced by it.

File: app/services/kubejob_service.py (hunk 3 above)
Comment (matthewberry) [reply app/services/kubejob_service.py]: Confirmed, and still live on `main` — fixed in #128.

One mechanism in the finding is wrong in a way that matters for the fix. `NoSuchBucket` is not how the missing-bucket case presents: a `stat_object` 404 has no body, so minio 7.1.17 falls through to `error_map[404]`, which returns **`NoSuchKey`** whenever `object_name` is set (`minio/api.py:361`). It is indistinguishable from a genuinely absent marker, so "treat a non-`NoSuchKey` `S3Error` as already sent" would not close that hole. `mark_email_as_sent` cannot recover either — `put_object` does not create a bucket.

Connection-refused is also not a path: that raises a urllib3 error, not an `S3Error`, so it escapes `check_file_exists` entirely and is caught by the sweep's per-job handler, sending nothing. Same for a 5xx, which raises `ServerError` (not an `S3Error` subclass). `AccessDenied` is the one case the `else` branch really does swallow.

So the reliable guard is the phase transition, not the marker. #128 moves the call inside `if phase_changed:`.

**Comment by matthewberry** [comment]
Closing this as superseded. Every change on this branch is already on `main`, landed independently — mostly through #121 (`fair/10-watcher-phase-fix`), which is why this PR now reports `CONFLICTING`.

| change on this branch | where it lives on `main` |
| --- | --- |
| DB row written before the Kubernetes Job (`created_new_job`) | `app/routers/job.py:127` |
| `_derive_phase` reading the whole condition list | `app/services/kubejob_service.py:252` |
| `_reconcile_job_phase` + on-connect sweep, `timeout_seconds = 600` | `app/services/kubejob_service.py:282`, `393`, `414` |
| per-event / per-job exception isolation | `app/services/kubejob_service.py:414` |
| `getattr(e, 'status', None) == 410` | `app/services/kubejob_service.py:427` |
| `watcher.run()` removed from the lifespan | `app/main.py` |
| `echo` following `DEBUG` instead of pinned `True` | `app/models/sqlmodel/db.py` |

`main` also goes further in places this branch did not: the router checks `create_job`'s returned `status`, the duplicate-submission path no longer forks a second Kubernetes Job or echoes back the owner's email, and `_reconcile_job_phase` refuses to resurrect a `canceled` job.

## Disposition of the review findings

**Already fixed on `main`:**

- **1 (HIGH), `create_job()`'s return value ignored** — correct about this branch. `main` checks `create_response.get('status') == STATUS_ERROR` at `app/routers/job.py:197` and calls `_mark_job_failed`, and `tests/characterization/test_jobs_api.py::TestKubernetesRejection` pins it.
- **4 (LOW), the `echo` change** — `main` uses `echo=DEBUG` imported from `config`, so the case-sensitivity divergence is gone. The other half stands and is a deployment decision, not a code one: `app/cfg/config.yaml:3` is `DEBUG` and no env values file overrides `server.loglevel`, so `echo` is on everywhere and SQLAlchemy logs every statement and parameter set in production.
- **5 (LOW), `self.logger.setLevel('DEBUG')`** — accurate about the branch, but the provenance is wrong: that is commit `088648e`, "turn up logging for only KubeWatcher", a deliberate change by @bodom0015, not a leftover of mine. `main` has it commented out again.
- **6 (LOW), the unreachable duplicate row constructor** — the reachability argument holds, and `main` deleted the block in the duplicate-submission rewrite.

**Real, and live on `main` — opened as #128:**

- **2 (HIGH), duplicate notification-email amplification.** One mechanism in the finding is wrong in a way that changes the fix; details in the thread above.
- **7 (LOW), `time_end` unset on the submission-failure path.**

**Rejected:**

- **3 (MEDIUM), the sweep clobbering ChemScraper's `completed`.** The premise does not hold; details in the thread above.

**Comment by matthewberry** [comment]
Correction to my closing comment above, on finding 4.

I wrote that `app/cfg/config.yaml:3` is `DEBUG` with no values file overriding `server.loglevel`, so `echo` is on in every deployed environment and the "turn down sqlalchemy transactions" change is a no-op as shipped. That is wrong.

I ran the grep on this branch, which is based on an old `main` and still carries `loglevel: "DEBUG"` at `chart/values.yaml:84`. Current `main` sets `chart/values.yaml:120: loglevel: "INFO"` — changed in `cee8751` — and that is what deployments use. `templates/configmap.yaml` renders `.Values.config` and `deployment.yaml:68-71` mounts it over `/code/app/cfg/config.yaml`, so the image's fallback never applies in a cluster. Nothing sets the `LOGLEVEL` env var `config.py:57` reads.

So `SERVER_LOGLEVEL` is `INFO`, `DEBUG = SERVER_LOGLEVEL.lower() == 'debug'` is False, and `create_async_engine(..., echo=DEBUG)` gets `echo=False`. @bodom0015's commit does exactly what it set out to do; SQLAlchemy is not logging every statement and parameter set in production, and there is nothing to action here.

Finding 4's other half still stands: deriving the flag case-sensitively in `db.py` rather than importing `DEBUG` from `config` was a real inconsistency, and `main` already fixed it. Nothing else in the disposition above depends on the loglevel reading, and no code was shipped on it — the change I wrote against it was on this branch and went away when the PR was closed as superseded.

Caught by review on #131.
